core/vote/vote_manager.go

```diff
@@ -0,0 +1,219 @@
+package vote
+
+import (
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/consensus"
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/eth/downloader"
+	"github.com/ethereum/go-ethereum/event"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/metrics"
+	"github.com/ethereum/go-ethereum/params"
+)
+
+// VoteManager will handle the vote produced by self.
+type VoteManager struct {
+	mux *event.TypeMux
+
+	chain       *core.BlockChain
+	chainconfig *params.ChainConfig
+
+	chainHeadCh  chan core.ChainHeadEvent
+	chainHeadSub event.Subscription
+
+	pool    *VotePool
+	signer  *VoteSigner
+	journal *VoteJournal
+
+	engine consensus.PoSA
+}
+
+func NewVoteManager(mux *event.TypeMux, chainconfig *params.ChainConfig, chain *core.BlockChain, pool *VotePool, journalPath, blsPasswordPath, blsWalletPath string, engine consensus.PoSA) (*VoteManager, error) {
+	voteManager := &VoteManager{
+		mux: mux,
+
+		chain:       chain,
+		chainconfig: chainconfig,
+		chainHeadCh: make(chan core.ChainHeadEvent, chainHeadChanSize),
+
+		pool:   pool,
+		engine: engine,
+	}
+
+	// Create voteSigner.
+	voteSigner, err := NewVoteSigner(blsPasswordPath, blsWalletPath)
+	if err != nil {
+		return nil, err
+	}
+	log.Info("Create voteSigner successfully")
+	voteManager.signer = voteSigner
+
+	// Create voteJournal
+	voteJournal, err := NewVoteJournal(journalPath)
+	if err != nil {
+		return nil, err
+	}
+	log.Info("Create voteJournal successfully")
+	voteManager.journal = voteJournal
+
+	// Subscribe to chain head event.
+	voteManager.chainHeadSub = voteManager.chain.SubscribeChainHeadEvent(voteManager.chainHeadCh)
```

Comment:
since we use a lot of `Subscribe` here, please take care of graceful shutdown, incase send object a closed channel or read from a closed channel, which may cause panic.

Reply:
I will review this  theme  related